contrib/aws/aws-sdk-go/aws/aws.go

```diff
@@ -0,0 +1,93 @@
+package aws
+
+import (
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+type handlers struct {
+	cfg *config
+}
+
+// WrapSession wraps an aws Session so that requests/responses are traced.
+func WrapSession(s *session.Session, opts ...Option) *session.Session {
+	cfg := new(config)
+	defaults(cfg)
+	for _, opt := range opts {
+		opt(cfg)
+	}
+	h := &handlers{cfg: cfg}
+	s = s.Copy()
+	s.Handlers.Send.PushFrontNamed(request.NamedHandler{
+		Name: "gopkg.in/DataDog/dd-trace-go.v1/contrib/aws/aws-sdk-go/aws/handlers.Send",
+		Fn:   h.Send,
+	})
+	s.Handlers.Complete.PushBackNamed(request.NamedHandler{
+		Name: "gopkg.in/DataDog/dd-trace-go.v1/contrib/aws/aws-sdk-go/aws/handlers.Complete",
+		Fn:   h.Complete,
+	})
+	return s
+}
+
+func (h *handlers) Send(req *request.Request) {
+	_, ctx := tracer.StartSpanFromContext(req.Context(), h.operationName(req),
+		tracer.SpanType(ext.AppTypeHTTP),
+		tracer.ServiceName(h.serviceName(req)),
+		tracer.ResourceName(h.resourceName(req)),
+		tracer.Tag(awsAgentTag, h.awsAgent(req)),
+		tracer.Tag(awsOperationTag, h.awsOperation(req)),
+		tracer.Tag(awsRegionTag, h.awsRegion(req)),
+		tracer.Tag(ext.HTTPMethod, req.Operation.HTTPMethod),
+		tracer.Tag(ext.HTTPURL, req.HTTPRequest.URL.String()),
+	)
+	req.SetContext(ctx)
+}
+
+func (h *handlers) Complete(req *request.Request) {
+	span, ok := tracer.SpanFromContext(req.Context())
+	if !ok {
+		return
+	}
+	if req.HTTPResponse != nil {
+		span.SetTag(ext.HTTPCode, strconv.Itoa(req.HTTPResponse.StatusCode))
+	}
+	span.Finish(tracer.WithError(req.Error))
```

Comment:
Should be ok in this case since the field has an `error` type. The problem arises when a pointer to a struct that happens to implement the error interface is used, since 

```go
var x *someerrortype
if somecondition {
    x = &someerrortype{...}
}
```

ends up being set to a nil pointer, whereas

```go
var x error
if somecondition {
    x = &someerrortype{...}
}
```

ends up being a nil interface

Reply:
Right, I guess it can happen anywhere. Errors having nil pointer values seems like a bad practice anyway.